ZOOKEEPER-4289: Reduce the performance impact of Prometheus metrics

li4wang opened the pull request:
Enabling Prometheus provider has significant impact on the performance of both read and write operations. This is to reduce the impact by making the Prometheus summary reporting as an async operation.

Load test results showed that the avg read latency and throughput after the fix is on par with the performance when Prometheus is disabled. For writes, the avg latency was reduced 25% and the avg throughput was increased 20% after the fix.

eolivelli commented on zookeeper-metrics-providers/zookeeper-prometheus-metrics/src/main/java/org/apache/zookeeper/metrics/prometheus/PrometheusMetricsProvider.java:
```diff
@@ -70,6 +99,11 @@
     private Server server;
     private final MetricsServletImpl servlet = new MetricsServletImpl();
     private final Context rootContext = new Context();
+    private final int numWorkerThreads = Integer.getInteger(ZOOKEEPER_PROMETHEUS_NUM_WORKER_THREADS,
```
we should read these configuration parameters in the "configure" method
this way we are not depending on system properties

li4wang replied:
I see. I will make the change so they will be read from configuration file just the other properties such as `httpHost`

We also need to document them in the zookeeperAdmin.md, right?

eolivelli reviewed (changes requested):
Great work !

I left a comment about configuration, apart from that it looks great to me

eolivelli commented:
@merlimat @dlg99 @lhotari 
you may be interested in this work

li4wang commented:
Thanks a lot for comments and feedback. @eolivelli

li4wang commented:
Hi @eolivelli, thanks again for your feedback. I have changed the way of reading the config parameters and also added the documentation for the properties. Would you mind taking a look at it?

anmolnar reviewed (approved):
Neat. Does it make sense to ever disable this feature? (Setting num_workers to 0)

li4wang commented:
Thanks a lot for reviewing it, @anmolnar .

No, I think we would want to enable this in normal cases. I added the support for num_works to 0 to provide the flexibility /possibility to switch between sync vs async for the cases of

1.  people want to run benchmark comparison tests with and without the fix
2. people really have limited number of threads they can use for the system

I can definitely remove the support if we don't need to support the above the cases.

eolivelli reviewed (approved):
LGTM

eolivelli commented:
we should cherry pick this patch to 3.7.1 and probably also to 3.6.4.

li4wang commented:
Hi @eolivelli thanks for taking time review it.
> we should cherry pick this patch to 3.7.1 and probably also to 3.6.4.
What's the process of cherry picking changes? Do I need to open two separate JIRA tickets and having PR for each branch?

eolivelli commented:
No need  for a new jira. We can set multiple fix version.

If the change applies cleanly to the other branches there  is no need for additional prs

li4wang commented:
Excellent! Thanks @eolivelli 

Please let me know if anything else needed from me for getting the changes to the master and other branches.  This is the first time for me submitting changes, so I am not quite sure the process.

eolivelli commented:
@li4wang sure.
I will try to make CI pass if it is a case of flaky tests.
I will try to cherry pick to branch-3.7 and branch 3.6. I will ping you in case of problems

li4wang commented:
@eolivelli is there anything outstanding that I can help to get the changes merged so we can get the changes from the Zookeeper repo and prepare for 3.7 rollout?

eolivelli commented:
@li4wang I have committed the patch to master branch.
It does not apply to branch-3.7.

Do you mind sending another patch for branch-3.7 ?

li4wang commented:
Thanks for committing the patch to the master branch, Enrico.

I will send another one for branch-3.7.

Cheers,

Li

On Wed, Jul 28, 2021 at 2:27 AM Enrico Olivelli ***@***.***>
wrote:

> @li4wang <https://github.com/li4wang> I have committed the patch to
> master branch.
> It does not apply to branch-3.7.
>
> Do you mind sending another patch for branch-3.7 ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/zookeeper/pull/1698#issuecomment-888158310>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQMZS2C4MP7Q5QVZWEQTT6TTZ7EPTANCNFSM45DO6GEA>
> .
>

ChaosJu commented:
great work
